vstart_runner: accepts multiple commands as str too

rishabh-d-dave opened the pull request:
Teuthology's orchestra.remote.run() can execute multiple commands passed
to it in form of character string where commands are separated by
newlines. Let's add that ability to vstart_runner.LocalRemote.run() as
well.
    
Fixes: https://tracker.ceph.com/issues/47849

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Link to laat API job - https://jenkins.ceph.com/job/ceph-api/5023/

rishabh-d-dave commented:
Link to laat API job - https://jenkins.ceph.com/job/ceph-api/5023/

rishabh-d-dave commented:
The persistent error on Ceph API tests (see https://jenkins.ceph.com/job/ceph-api/6273/) is due to a bug in `mgr_test_case.py`. Here's the fix for that bug - https://github.com/ceph/ceph/pull/37992.

Marking this as DNM until that PR merges.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
With the fix for dashboard QA, the Ceph API tests passed - https://jenkins.ceph.com/job/ceph-api/6374/

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Following is the part of logs where the failure occurred -
```
2020-11-18 08:23:30,396.396 INFO:__main__:Starting test: test_module_commands (tasks.mgr.test_module_selftest.TestModuleSelftest)
2020-11-18 08:23:30,396.396 INFO:__main__:> ./bin/ceph log Starting test tasks.mgr.test_module_selftest.TestModuleSelftest.test_module_commands
2020-11-18 08:23:31,402.402 INFO:__main__:> ps ww -u1108
2020-11-18 08:23:31,432.432 INFO:__main__:Found ps line for daemon: 25255 ?        Ssl    0:05 ./bin/./ceph-mgr -i x
2020-11-18 08:23:31,433.433 INFO:__main__:> ps ww -u1108
2020-11-18 08:23:31,464.464 INFO:__main__:Found ps line for daemon: 25255 ?        Ssl    0:05 ./bin/./ceph-mgr -i x
2020-11-18 08:23:31,465.465 INFO:__main__:Killing PID 25255 for mgr.x
2020-11-18 08:23:31,465.465 INFO:__main__:> ps ww -u1108
2020-11-18 08:23:31,495.495 INFO:__main__:No match for mgr x
2020-11-18 08:23:31,496.496 INFO:__main__:> ps ww -u1108
2020-11-18 08:23:31,520.520 INFO:__main__:No match for mgr x
2020-11-18 08:23:31,521.521 INFO:__main__:> ./bin/ceph mgr fail x
2020-11-18 08:23:33,038.038 INFO:__main__:> ./bin/ceph mgr module ls
2020-11-18 08:23:33,705.705 INFO:__main__:> ./bin/ceph mgr module disable iostat
2020-11-18 08:23:34,434.434 INFO:__main__:> ./bin/ceph mgr module disable selftest
2020-11-18 08:23:35,475.475 INFO:__main__:> ps ww -u1108
2020-11-18 08:23:35,504.504 INFO:__main__:No match for mgr x
2020-11-18 08:23:35,505.505 INFO:__main__:> ./bin/./ceph-mgr -i x
2020-11-18 08:23:35,634.634 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:41,292.292 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:41,955.955 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:42,615.615 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:43,275.275 INFO:__main__:> ./bin/ceph mgr module ls
2020-11-18 08:23:43,925.925 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:44,593.593 INFO:__main__:> ./bin/ceph mgr metadata
2020-11-18 08:23:45,218.218 INFO:tasks.mgr.mgr_test_case:Loading Mgr module selftest ...
2020-11-18 08:23:45,219.219 INFO:__main__:> ./bin/ceph mgr module enable selftest --force
2020-11-18 08:23:46,684.684 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:52,349.349 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:58,017.017 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-18 08:23:58,678.678 INFO:tasks.mgr.mgr_test_case:Restarted after module load (new active x/5960)
2020-11-18 08:23:58,680.680 INFO:__main__:> ./bin/ceph mgr module disable selftest
2020-11-18 08:24:00,193.193 INFO:__main__:> ./bin/ceph mgr self-test run
2020-11-18T08:24:00.427+0000 7f010a95f700 -1 WARNING: all dangerous and experimental features are enabled.
2020-11-18T08:24:00.483+0000 7f010a95f700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOTSUP: Module 'selftest' is not enabled (required by command 
```

Since locally this test passed for this branch, I am running API tests once again.

Link to the previous run - https://jenkins.ceph.com/job/ceph-api/6994/

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Strange, I see "Took 0 ms on master" on API test job - https://jenkins.ceph.com/job/ceph-api/7040/. Retriggering...

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Seeing the same failure again -

```
2020-11-19 07:02:52,275.275 INFO:__main__:Starting test: test_module_commands (tasks.mgr.test_module_selftest.TestModuleSelftest)
2020-11-19 07:02:52,275.275 INFO:__main__:> ./bin/ceph log Starting test tasks.mgr.test_module_selftest.TestModuleSelftest.test_module_commands
2020-11-19 07:02:53,305.305 INFO:__main__:> ps ww -u1001
2020-11-19 07:02:53,319.319 INFO:__main__:Found ps line for daemon: 30419 ?        Ssl    0:04 ./bin/./ceph-mgr -i x
2020-11-19 07:02:53,319.319 INFO:__main__:> ps ww -u1001
2020-11-19 07:02:53,337.337 INFO:__main__:Found ps line for daemon: 30419 ?        Ssl    0:04 ./bin/./ceph-mgr -i x
2020-11-19 07:02:53,337.337 INFO:__main__:Killing PID 30419 for mgr.x
2020-11-19 07:02:53,338.338 INFO:__main__:> ps ww -u1001
2020-11-19 07:02:53,350.350 INFO:__main__:No match for mgr x
2020-11-19 07:02:53,351.351 INFO:__main__:> ps ww -u1001
2020-11-19 07:02:53,369.369 INFO:__main__:No match for mgr x
2020-11-19 07:02:53,370.370 INFO:__main__:> ./bin/ceph mgr fail x
2020-11-19 07:02:54,359.359 INFO:__main__:> ./bin/ceph mgr module ls
2020-11-19 07:02:54,922.922 INFO:__main__:> ./bin/ceph mgr module disable iostat
2020-11-19 07:02:56,371.371 INFO:__main__:> ./bin/ceph mgr module disable selftest
2020-11-19 07:02:57,409.409 INFO:__main__:> ps ww -u1001
2020-11-19 07:02:57,430.430 INFO:__main__:No match for mgr x
2020-11-19 07:02:57,431.431 INFO:__main__:> ./bin/./ceph-mgr -i x
2020-11-19 07:02:57,528.528 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-19 07:03:03,076.076 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-19 07:03:03,625.625 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-19 07:03:04,175.175 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-19 07:03:04,732.732 INFO:__main__:> ./bin/ceph mgr module ls
2020-11-19 07:03:05,289.289 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-19 07:03:05,836.836 INFO:__main__:> ./bin/ceph mgr metadata
2020-11-19 07:03:06,395.395 INFO:tasks.mgr.mgr_test_case:Loading Mgr module selftest ...
2020-11-19 07:03:06,396.396 INFO:__main__:> ./bin/ceph mgr module enable selftest --force
2020-11-19 07:03:07,388.388 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-19 07:03:12,945.945 INFO:__main__:> ./bin/ceph mgr dump --format=json-pretty
2020-11-19 07:03:13,516.516 INFO:tasks.mgr.mgr_test_case:Restarted after module load (new active x/5935)
2020-11-19 07:03:13,517.517 INFO:__main__:> ./bin/ceph mgr module disable selftest
2020-11-19 07:03:14,321.321 INFO:__main__:> ./bin/ceph mgr self-test run
2020-11-19T07:03:14.499+0000 7f318add9700 -1 WARNING: all dangerous and experimental features are enabled.
2020-11-19T07:03:14.531+0000 7f318add9700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOTSUP: Module 'selftest' is not enabled (required by command 'mgr self-test run'): use `ceph mgr module enable selftest` to enable it
```

lxbsz commented on qa/tasks/vstart_runner.py:
```diff
@@ -388,10 +383,41 @@ def _perform_checks_and_return_list_of_args(self, args, omit_sudo):
+        if isinstance(args, str):
+            if '\n' in args:
+                args = args.split('\n')
+                args = [a.split() for a in args]
```
BTW, will the following case happen ?

If the `args` is something like:  `"ls /mnt/cpehfs '\n' cd /mnt/cephfs/"`, after this it will be: `['ls', '/mnt/cephfs', 'cd', '/mnt/cephfs/']`.

Should we add a `;` between the two commands? like:
`['ls', '/mnt/cephfs', ';', 'cd', '/mnt/cephfs/']` or `['ls', '/mnt/cephfs', '&&', 'cd', '/mnt/cephfs/']`.

rishabh-d-dave replied:
I don't think that would work. Python's shell simulation would think of the latter elements as arguments for the first element.

lxbsz replied:
BTW, do you have the use case of it ?

rishabh-d-dave replied:
Just as I had expected, `;` is not interpreted as command separator -
```
2020-12-01T19:42:27.929 INFO:__main__:> ls ; ls
2020-12-01T19:42:27.942 INFO:__main__:> sudo umount /tmp/tmpkuz9qrta/mnt.0
2020-12-01T19:42:27.986 INFO:__main__:> rmdir -- /tmp/tmpkuz9qrta/mnt.0
2020-12-01T19:42:27.990 INFO:__main__:> ./bin/ceph config rm mon mon_allow_pool_delete
2020-12-01T19:42:28.368 INFO:__main__:> ./bin/ceph log Ended test tasks.cephfs.test_cephfs_shell.TestCD.test_cd_with_no_args
2020-12-01T19:42:29.497 INFO:__main__:Stopped test: test_cd_with_no_args (tasks.cephfs.test_cephfs_shell.TestCD)
Test that when cd is issued without any arguments, CWD is changed in 75.512632s
2020-12-01T19:42:29.497 INFO:__main__:test_cd_with_no_args (tasks.cephfs.test_cephfs_shell.TestCD)
2020-12-01T19:42:29.497 INFO:__main__:Test that when cd is issued without any arguments, CWD is changed ... ERROR
2020-12-01T19:42:29.497 INFO:__main__:
2020-12-01T19:42:29.498 INFO:__main__:======================================================================
2020-12-01T19:42:29.498 INFO:__main__:ERROR: test_cd_with_no_args (tasks.cephfs.test_cephfs_shell.TestCD)
2020-12-01T19:42:29.498 INFO:__main__:Test that when cd is issued without any arguments, CWD is changed
2020-12-01T19:42:29.498 INFO:__main__:
2020-12-01T19:42:29.498 INFO:__main__:----------------------------------------------------------------------
2020-12-01T19:42:29.498 INFO:__main__:Traceback (most recent call last):
2020-12-01T19:42:29.498 INFO:__main__:  File "/home/rishabh/repos/ceph/minor/qa/tasks/cephfs/test_cephfs_shell.py", line 495, in test_cd_with_no_args
2020-12-01T19:42:29.498 INFO:__main__:    self.mount_a.client_remote.run(args='ls\nls')
2020-12-01T19:42:29.498 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 423, in run
2020-12-01T19:42:29.498 INFO:__main__:    return self._do_run(**kwargs)
2020-12-01T19:42:29.498 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 492, in _do_run
2020-12-01T19:42:29.498 INFO:__main__:    proc.wait()
2020-12-01T19:42:29.498 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 208, in wait
2020-12-01T19:42:29.498 INFO:__main__:    raise CommandFailedError(self.args, self.exitstatus)
2020-12-01T19:42:29.498 INFO:__main__:teuthology.exceptions.CommandFailedError: Command failed with status 2: ['ls', ';', 'ls']
```

Return value 2 stands `No such file or directory`.

rishabh-d-dave replied:
> BTW, do you have the use case of it ?

Use case for `;` is not interpreted as separator or something else?

rishabh-d-dave replied:
@lxbsz 
Ah, I see misunderstanding  here...

> If the args is something like: "ls /mnt/cpehfs '\n' cd /mnt/cephfs/", after this it will be: ['ls', '/mnt/cephfs', 'cd', '/mnt/cephfs/'].


Command like "ls /mnt/cpehfs '\n' cd /mnt/cephfs/" would be split into [['ls', '/mnt/cephfs'], ['cd', '/mnt/cephfs/']].


```
if '\n' in args: # if args = 'ls /mnt/cpehfs '\n' cd /mnt/cephfs/'
                args = args.split('\n') # args would become ['ls /mnt/cpehfs', 'cd /mnt/cephfs/']
                args = [a.split() for a in args] # and now it would become [['ls', '/mnt/cephfs'], ['cd', '/mnt/cephfs/']]
```

lxbsz replied:
Okay, that looks good.

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Moved one of the two commits to a PR #38443. Although the commit here doesn't directly depend on the commits on PR #38443, full resolution of the tracker issue[[1]]( https://tracker.ceph.com/issues/47849) does depend on it since teuthology.orchestra. RemoteProcess.write_file() can't be run without setting `shell` to `True`.

rishabh-d-dave commented:
@lxbsz Can you please finish reviewing this PR too? It will rebase soon, but otherwise this PR is final. Thanks again!

lxbsz reviewed (approved):
LGTM.

batrick commented:
@rishabh-d-dave status on this? Still DNM?

rishabh-d-dave commented:
@batrick The PR itself is ready for merge but fixing https://tracker.ceph.com/issues/47849 requires PR #38443 to merge first.

https://github.com/ceph/ceph/pull/37655#issuecomment-739798063

lxbsz reviewed (approved):
LGTM.

tchaikov commented:
jenkins test make check

rishabh-d-dave commented:
@batrick Rebased this PR and retested locally. This one's ready for merge. https://gist.github.com/rishabh-d-dave/16697693b4cab3a83b4190dad5394cfb

lxbsz reviewed (approved):
LGTM.

rishabh-d-dave commented:
@batrick PTAL.

rishabh-d-dave commented:
@batrick I retested the PR and copied the output here - https://gist.github.com/rishabh-d-dave/16697693b4cab3a83b4190dad5394cfb

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test make check

tchaikov commented on qa/tasks/vstart_runner.py:
```diff
@@ -405,10 +400,41 @@ def _perform_checks_and_return_list_of_args(self, args, omit_sudo):
+        We might have multiple commands in kwargs['args'] as list within list
+        or as string with newlines.
+        """
+        cmds = type(self).split_args(kwargs['args'])
```
when performing method resolution for `instance.func()`, in addition to regular functions, both the static methods and class methods are considered. see https://docs.python.org/3/library/functions.html#staticmethod

so no need to call `split_args()` this way, could just put
```py
cmds = self.split(kwargs['args']
```

rishabh-d-dave replied:
True, I've made the change. Thanks!

tchaikov commented on qa/tasks/vstart_runner.py:
```diff
@@ -405,10 +400,41 @@ def _perform_checks_and_return_list_of_args(self, args, omit_sudo):
+                args = [a.split() for a in args]
+            else:
+                args = args.split()
+        return args
```
i'd suggest normalize the `args` here instead of returning different instances of different types, and let the caller to dispatch by types. this is error prune, and difficult to maintain. could just put
```py
if isinstance(args, str):
    return [cmd.split() for cmd in args.split]
else:
    assert(isinstance(args, list))
    assert(all(isinstance(arg, str) for arg in args))
    return [args]
```

rishabh-d-dave replied:
Yes, keeping a single return value type would be better. I've made the change.

tchaikov commented:
jenkins test make check

tchaikov commented on qa/tasks/vstart_runner.py:
```diff
@@ -405,10 +400,46 @@ def _perform_checks_and_return_list_of_args(self, args, omit_sudo):
+
+        # guard against empty list. 'ls /usr\n' will produce list containing
+        # empty list, for example.
+        while [] in cmds:
+            cmds.remove([])
+
+        return cmds
```
```suggestion
       return [cmd for cmd in cmds if cmd]
```

tchaikov commented:
- https://pulpito.ceph.com/kchai-2021-07-01_14:09:56-rados-wip-kefu-testing-2021-07-01-1912-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-02_03:53:50-rados-wip-kefu-testing-2021-07-02-0958-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-02_12:31:26-rados-wip-kefu-testing-2021-07-02-1417-distro-basic-smithi/

failures are not related to this changeset.

rishabh-d-dave commented:
Thanks for testing @tchaikov!

I am marking this PR DNM to block merge, Patrick wants some more changes.

batrick commented on qa/tasks/vstart_runner.py:
```diff
@@ -434,8 +428,39 @@ def _do_run(self, args, check_status=True, wait=True, stdout=None,
+                stderr=None, cwd=None, stdin=None, logger=None, label=None,
+                env=None, timeout=None, omit_sudo=True, shell=True):
+        args = self._perform_checks(args, omit_sudo, shell)
 
```
```suggestion
prefix = """
adjust-ulimits() {
  "$@"
}
ceph-coverage() {
  "$@"
}
"""
cmd = prefix + quote(args)
```

rishabh-d-dave replied:
IMO this makes the log/output much less readable, compact and pretty -

```
}
ceph-coverage() {
  "$@"
}
./bin/ceph osd pool rm cephfs_data cephfs_data --yes-i-really-really-mean-it
2021-07-15T03:05:06.036 DEBUG:__main__:>
adjust-ulimits() {
  "$@"
}
ceph-coverage() {
  "$@"
}
./bin/ceph config rm mon mon_allow_pool_delete
2021-07-15T03:05:06.514 DEBUG:__main__:>
adjust-ulimits() {
  "$@"
}
ceph-coverage() {
  "$@"
}
```

I tried reducing number of lines it consumes but my opinion didn't change -

```
2021-07-15T03:06:00.323 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
./bin/ceph osd pool rm cephfs_metadata cephfs_metadata --yes-i-really-really-mean-it
2021-07-15T03:06:01.384 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
./bin/ceph osd pool rm cephfs_data cephfs_data --yes-i-really-really-mean-it
2021-07-15T03:06:02.450 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
./bin/ceph config rm mon mon_allow_pool_delete
2021-07-15T03:06:02.897 DEBUG:__main__:> adjust-ulimits() { "$@"; }
ceph-coverage() { "$@"; }
```

Deleting these commands is best option IMO. We can add warning for deletion of those commands at the very beginning of vstart_runner.py's output, in docstrings and add a warning in the documentation as well - https://docs.ceph.com/en/latest/dev/developer_guide/running-tests-locally/#running-test-using-vstart-runner-py.

batrick replied:
There's  no reason you can prepend those function hooks prior to invoking `/bin/sh`? I don't see why it needs to end up in the log.

rishabh-d-dave replied:
Deleting these commands is best option IMO. We can add warning for deletion of those commands at the very beginning of vstart_runner.py's output, in docstrings and add a warning in the documentation as well - https://docs.ceph.com/en/latest/dev/developer_guide/running-tests-locally/#running-test-using-vstart-runner-py.

batrick commented on qa/tasks/vstart_runner.py:
```diff
@@ -434,8 +428,39 @@ def _do_run(self, args, check_status=True, wait=True, stdout=None,
+        args = self._perform_checks(args, omit_sudo, shell)
 
         if shell:
             subproc = subprocess.Popen(quote(args),
```
```suggestion
            subproc = subprocess.Popen(cmd,
```

rishabh-d-dave replied:
Done.

batrick commented on qa/tasks/vstart_runner.py:
```diff
@@ -434,8 +428,39 @@ def _do_run(self, args, check_status=True, wait=True, stdout=None,
+        return args
+
+    def _perform_checks(self, args, omit_sudo, shell):
+        if isinstance(args, str):
```
Convert list to string so we don't need to duplicate the checks.

rishabh-d-dave replied:
Done.

batrick commented on qa/tasks/vstart_runner.py:
```diff
@@ -361,12 +361,25 @@ def sudo_write_file(self, path, data, **kwargs):
+            if omit_sudo:
+                args.replace('sudo', '')
 
+        args = args.replace('adjust-ulimits','')
+        args = args.replace('ceph-coverage', '')
```
Let's remove these and just override the commands in the shell script we run (comment later).

rishabh-d-dave replied:
Replied on second thread of this conversation, leave that one open and closing this one.

rishabh-d-dave commented:
@batrick I've incorporated changes requested by you and I've updated docs and docstrings to add the warnings I talked [here](https://github.com/ceph/ceph/pull/37655#discussion_r681193272).

@tchaikov Please re-review this PR. The current idea is to convert all command arguments to `str` from `list` and then execute them by setting Popen.shell to `True`.


@ceph/dashboard @ceph/api This PR heavily affects the code that issues the commands. Please take a look since most probably affect Ceph API  tests as well.

rishabh-d-dave commented:
jenkins test make check arm64

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-pull-requests/80413/
https://jenkins.ceph.com/job/ceph-pull-requests-arm64/15363/
https://jenkins.ceph.com/job/ceph-api/22501/

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test make check arm64

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-api/22569/consoleFull#527393977dc3600c-28d8-42e2-9653-291a0c275fdf


```
2021-08-03 15:11:57,391.391 INFO:__main__:Use the selftest module to test the cluster/audit log interface. ... ERROR
2021-08-03 15:11:57,392.392 INFO:__main__:
2021-08-03 15:11:57,392.392 INFO:__main__:
2021-08-03 15:11:57,392.392 INFO:__main__:======================================================================
2021-08-03 15:11:57,392.392 INFO:__main__:ERROR: test_selftest_cluster_log (tasks.mgr.test_module_selftest.TestModuleSelftest)
2021-08-03 15:11:57,392.392 INFO:__main__:Use the selftest module to test the cluster/audit log interface.
2021-08-03 15:11:57,392.392 INFO:__main__:----------------------------------------------------------------------
2021-08-03 15:11:57,392.392 INFO:__main__:Traceback (most recent call last):
2021-08-03 15:11:57,392.392 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/test_module_selftest.py", line 237, in test_selftest_cluster_log
2021-08-03 15:11:57,392.392 INFO:__main__:    with self.assert_cluster_log(log_message):
2021-08-03 15:11:57,393.393 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/ceph_test_case.py", line 115, in __enter__
2021-08-03 15:11:57,393.393 INFO:__main__:    self.watcher_process = ceph_manager.run_ceph_w(watch_channel,
2021-08-03 15:11:57,393.393 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 801, in run_ceph_w
2021-08-03 15:11:57,393.393 INFO:__main__:    proc = self.controller.run(args=args, wait=False, stdout=StringIO(),
2021-08-03 15:11:57,393.393 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 436, in run
2021-08-03 15:11:57,393.393 INFO:__main__:    return self._do_run(**kwargs)
2021-08-03 15:11:57,393.393 INFO:__main__:  File "../qa/tasks/vstart_runner.py", line 446, in _do_run
2021-08-03 15:11:57,393.393 INFO:__main__:    subproc = subprocess.Popen(args,
2021-08-03 15:11:57,393.393 INFO:__main__:  File "/usr/lib/python3.8/subprocess.py", line 854, in __init__
2021-08-03 15:11:57,393.393 INFO:__main__:    if self.text_mode:
2021-08-03 15:11:57,393.393 INFO:__main__:  File "/usr/lib/python3.8/subprocess.py", line 1702, in _execute_child
2021-08-03 15:11:57,394.394 INFO:__main__:    if errno_num != 0:
2021-08-03 15:11:57,394.394 INFO:__main__:FileNotFoundError: [Errno 2] No such file or directory: './bin/ceph -w'
2021-08-03 15:11:57,394.394 INFO:__main__:
Cannot find device "ceph-brx"
2021-08-03 15:11:57,425.425 INFO:__main__:
2021-08-03 15:11:57,426.426 INFO:__main__:----------------------------------------------------------------------
2021-08-03 15:11:57,426.426 INFO:__main__:Ran 307 tests in 3539.873s
2021-08-03 15:11:57,426.426 INFO:__main__:
2021-08-03 15:11:57,426.426 INFO:__main__:FAILED (errors=1, skipped=1)
```

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-api/23154/

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test signed

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test make check

lxbsz commented on qa/tasks/vstart_runner.py:
```diff
@@ -603,6 +603,8 @@ def safe_kill(pid):
     os.kill annoyingly raises exception if process already dead.  Ignore it.
     """
     try:
+        # XXX: since os.kill()s fails in case of process launched with root
```
`s/os.kill()s/os.kill()/` or `s/fails/fail/` ?

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Link to Ceph API job - https://jenkins.ceph.com/job/ceph-api/35080/

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
Link to Ceph API CI job for run before last push - https://jenkins.ceph.com/job/ceph-api/35074/.

rishabh-d-dave commented:
Link to latest cephAPI CI job - https://jenkins.ceph.com/job/ceph-api/3508. Failure -

```
2022-04-17 16:34:36,374.374 INFO:__main__:Starting test: test_a_set_login_credentials (tasks.mgr.dashboard.test_auth.AuthTest)
2022-04-17T16:34:38.185+0000 7ffa86a5f700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-17T16:34:38.205+0000 7ffa86a5f700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOENT: User 'admin2' does not exist
2022-04-17T16:34:40.613+0000 7f80e44ba700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-17T16:34:40.633+0000 7f80e44ba700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOENT: User 'admin2' does not exist
```

rishabh-d-dave commented:
Ceph API CI job failed. Link - https://jenkins.ceph.com/job/ceph-api/35126/consoleFull#12945576334b0cfa0c-a892-49e0-a359-82544e1a192e. Following are problematic error messages I could find -
```
2022-04-18 11:11:49,513.513 INFO:__main__:Stopped test: test_version (tasks.mgr.dashboard.test_api.VersionReqTest) in 1.5585s
2022-04-18 11:12:32,556.556 INFO:__main__:Starting test: test_a_set_login_credentials (tasks.mgr.dashboard.test_auth.AuthTest)
2022-04-18T11:12:34.161+0000 7f7776be5700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-18T11:12:34.181+0000 7f7776be5700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOENT: User 'admin2' does not exist
2022-04-18T11:12:36.609+0000 7fd56e945700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-18T11:12:36.621+0000 7fd56e945700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOENT: User 'admin2' does not exist
2022-04-18 11:12:39,872.872 INFO:__main__:test_standby (tasks.mgr.test_dashboard.TestDashboard) ... ok
```

```
2022-04-18 11:13:13,597.597 INFO:__main__:Starting test: test_login_without_password (tasks.mgr.dashboard.test_auth.AuthTest)
2022-04-18T11:13:15.098+0000 7f1fb97aa700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-18T11:13:15.118+0000 7f1fb97aa700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOENT: User 'admin2' does not exist
2022-04-18T11:13:15.498+0000 7fabf1217700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-18T11:13:15.518+0000 7fabf1217700 -1 WARNING: all dangerous and experimental features are enabled.
Error EINVAL: Input file not specified: Please specify the file containing password/secret with "-i" option
2022-04-18 11:13:16,602.602 INFO:__main__:Stopped test: test_login_without_password (tasks.mgr.dashboard.test_auth.AuthTest) in 3.005544s
```

rishabh-d-dave commented:
Retriggering make check since error from previous make check job doesn't look releveant.

```
 31 - run-rbd-unit-tests-1.sh (Failed)
```

https://jenkins.ceph.com/job/ceph-pull-requests/93743/

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
```
2022-04-18 18:20:17,973.973 INFO:__main__:Starting test: test_module_commands (tasks.mgr.test_module_selftest.TestModuleSelftest)
That module-handled commands have appropriate  behavior on
2022-04-18T18:20:44.343+0000 7fbe9721e700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-18T18:20:44.359+0000 7fbe95fbc700 -1 WARNING: all dangerous and experimental features are enabled.
Error ENOTSUP: Module 'selftest' is not enabled (required by command 'mgr self-test run'): use `ceph mgr module enable selftest` to enable it
2022-04-18T18:20:51.792+0000 7f429a5b6700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-18T18:20:51.804+0000 7f4299354700 -1 WARNING: all dangerous and experimental features are enabled.
no valid command found; 10 closest matches:
osd perf
osd df [<output_method:plain|tree>] [<filter_by:class|name>] [<filter>]
osd blocked-by
osd pool stats [<pool_name>]
osd pool scrub <who>...
osd pool deep-scrub <who>...
osd pool repair <who>...
osd pool force-recovery <who>...
osd pool force-backfill <who>...
osd pool cancel-force-recovery <who>...
Error EINVAL: invalid command
2022-04-18T18:21:05.712+0000 7fc9329e7700 -1 WARNING: all dangerous and experimental features are enabled.
2022-04-18T18:21:05.728+0000 7fc931785700 -1 WARNING: all dangerous and experimental features are enabled.
Error EIO: Module 'selftest' has experienced an error and cannot handle commands: Synthetic exception in serve
2022-04-18 18:21:05,949.949 INFO:tasks.ceph_test_case:waiting 30s for health warning matching Module 'selftest' has failed: Synthetic exception in serve
2022-04-18 18:21:25,909.909 INFO:__main__:Stopped test: test_module_commands (tasks.mgr.test_module_selftest.TestModuleSelftest)
That module-handled commands have appropriate  behavior on in 67.935389s
2022-04-18 18:21:25,909.909 INFO:__main__:Starting test: test_module_remote (tasks.mgr.test_module_selftest.TestModuleSelftest)
```

vshankar reviewed (approved):
LGTM

rishabh-d-dave commented:
@vshankar @epuertat Recent push consists of no code changes. I've modified the commit message that was marked DNM and removed a comment from vstart_runner.py since it isn't relevant anymore.

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented on doc/dev/developer_guide/running-tests-locally.rst:
```diff
@@ -103,9 +103,10 @@ vstart_runner.py can take the following options -
-          commands with passwordless sudo access. A failure on the kernel
-          client may crash the host, so it's recommended to use this
-          functionality within a virtual machine.
+          commands with passwordless sudo access.
+
+.. note:: A failure on the kernel client may crash the host, so it's
+          recommended to use this functionality within a virtual machine.
```
These changes have been merged in a separate PR - https://github.com/ceph/ceph/pull/45923.

rishabh-d-dave commented:
Rebased and resolved a conflict in  qa/tasks/cephfs/mount.py.

ljflores commented:
@batrick does this PR look okay to you? I see you originally requested changes.

vshankar commented:
> @batrick does this PR look okay to you? I see you originally requested changes.

@batrick is on PTO. I ran this through fs suite and results look fine to me. I plan to merge this soon (post reabse).

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-May-04
